### PR TITLE
Frogger game

### PR DESCRIPTION
**Pull request by SamVal007**
Frogger game

[demo](https://samval007.github.io/frontend-nanodegree-arcade-game/)
[codebase](https://github.com/SamVal007/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/SamVal007/Frogger game task/app.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            this.y += playerStartValue.stepY;
+        }
+    }
+
+    plusScore() {
+        this.score++;
+    }
+
+    minusScore() {
+        this.score <= playerStartValue.startScore ? this.score = playerStartValue.startScore : this.score--;
```
Comment (OleksiyRudenko) [inline submissions/SamVal007/Frogger game task/app.js:104]: Using ternary operator as `if...else` is an anti-pattern.
```suggestion
        this.score = this.score <= playerStartValue.startScore ? playerStartValue.startScore : this.score - 1;
```
or
```suggestion
        if (this.score <= playerStartValue.startScore)
          this.score = playerStartValue.startScore;
        else
          this.score--;
```

File: submissions/SamVal007/Frogger game task/app.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        this.player = player;
+    }
+
+    update(dt) {
+        this.x += this.speed * dt;
+        if(this.x > enemyStartValue.maxX) {
+            this.x = enemyStartValue.minX;
+            this.speed = this.getRandomSpeed();
+        }
+        this.isCollision();
```
Comment (OleksiyRudenko) [inline submissions/SamVal007/Frogger game task/app.js:55]: By convention a function where name starts with `is`, `does` is read as a question and is expceted to return a boolean as a response. Then some actions are expected based on the value such function returns.

If a function encapsulates some activity then it is expected that its name describes the activity.

File: submissions/SamVal007/Frogger game task/app.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    isCollision() {
+        if(player.x < this.x + coordsGameGround.conflictX && player.x + coordsGameGround.conflictX > this.x && player.y < this.y + coordsGameGround.conflictY && coordsGameGround.conflictY + player.y > this.y) {
+            player.x = playerStartValue.x;
+            player.y = playerStartValue.y;
+            this.player.minusScore();
+        }
+    }
```
Comment (OleksiyRudenko) [inline submissions/SamVal007/Frogger game task/app.js:64]: This method refers to a global variable `player`. Classes shouldn't refer to global variables.

Although referring to global constants is ok. In this particular task referring to game engine resources is also OK. However, the global variable `player` is neither const nor engine's entity.

File: submissions/SamVal007/Frogger game task/app.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    plusScore() {
+        this.score++;
+    }
+
+    minusScore() {
+        this.score <= playerStartValue.startScore ? this.score = playerStartValue.startScore : this.score--;
+    }
+
+    prizePoint() {
```
Comment (OleksiyRudenko) [inline submissions/SamVal007/Frogger game task/app.js:107]: Method names should start with verbs.
Please give a somewhat better naming to `plusScore`, `minusScore` and `prizePoint`

File: submissions/SamVal007/Frogger game task/app.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    render() {
+        ctx.drawImage(Resources.get(this.sprite), this.x, this.y);
+    }
+}
+
+class Enemy extends Character {
+    constructor(x, y, sprite, player) {
+        super(x, y, sprite);
+        this.speed = this.getRandomSpeed();
+        this.player = player;
```
Comment (OleksiyRudenko) [inline submissions/SamVal007/Frogger game task/app.js:46]: `this.player` is used once only.
Consider this as a hint for a problem highlighted in a yet another comment here.

File: submissions/SamVal007/Frogger game task/app.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        38: 'up',
+        39: 'right',
+        40: 'down'
+    };
+
+    player.handleInput(allowedKeys[e.keyCode]);
+});
+
+const player = new Player(playerStartValue.x, playerStartValue.y, playerStartValue.icon);
+const allEnemies = [enemyStartValue.firstEnemy, enemyStartValue.secondEnemy, enemyStartValue.thirdEnemy].map(y =>  new Enemy(enemyStartValue.x, y, enemyStartValue.icon, player));
```
Comment (OleksiyRudenko) [inline submissions/SamVal007/Frogger game task/app.js:128]: If `enemyStartValue` contained a property (e.g. `initialY`) with an array of initial values then this fragment would be as short as `const allEnemies = enemyStartValue.initialY.map(y =>  new Enemy(enemyStartValue.x, y, enemyStartValue.icon, player));`

Whenever variables tend to contain numerical in their names like `first`, `second` etc this is a clear use case for an array. Imagine you had to create a dozen of enemies.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@SamVal007 well done!
However there is some room for improvement yet

File: submissions/SamVal007/Frogger game task/app.js (hunk 1 above)
Comment (SamVal007) [reply submissions/SamVal007/Frogger game task/app.js]: Ok, I understand this! Thank you!

File: submissions/SamVal007/Frogger game task/app.js (hunk 2 above)
Comment (SamVal007) [reply submissions/SamVal007/Frogger game task/app.js:55]: I agree, thank you!

File: submissions/SamVal007/Frogger game task/app.js (hunk 3 above)
Comment (SamVal007) [reply submissions/SamVal007/Frogger game task/app.js:64]: As I understand it  -we can use here just "player.minusScore();" without 'this'. Right?

**Comment by OleksiyRudenko** [comment]
@SamVal007 the last commit is irrelevant here. Please create a different PR for that different task and also exclude the last commit from this PR

**Comment by SamVal007** [comment]
> @SamVal007 the last commit is irrelevant here. Please create a different PR for that different task and also exclude the last commit from this PR

Yes,  i'm sorry didnt changed file here

File: submissions/SamVal007/Frogger game task/app.js (hunk 6 above)
Comment (SamVal007) [reply submissions/SamVal007/Frogger game task/app.js]: Thanks! That's really great idea!

File: submissions/SamVal007/Frogger game task/app.js (hunk 4 above)
Comment (SamVal007) [reply submissions/SamVal007/Frogger game task/app.js]: Ok, got it! 
Changed them to `increaseScore`, `decreaseScore`, `takingScorePoint`. I think, it's more understandable now

File: submissions/SamVal007/Frogger game task/app.js (hunk 5 above)
Comment (SamVal007) [reply submissions/SamVal007/Frogger game task/app.js:44]: I got it!

File: submissions/SamVal007/Frogger game task/app.js (hunk 3 above)
Comment (SamVal007) [reply submissions/SamVal007/Frogger game task/app.js]: I got it! Yeah, really i should put there ` this.player`

**Comment by OleksiyRudenko** [comment]
@SamVal007 still need to fix issue with irrelevant files in this PR. Feel free seking help in the chat

**Comment by SamVal007** [comment]
> @SamVal007 still need to fix issue with irrelevant files in this PR. Feel free seking help in the chat

**Comment by SamVal007** [comment]
Thank you!
Hopefully, I did it.

**Pull request merged**